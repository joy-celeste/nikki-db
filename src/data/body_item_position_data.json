{"position":{"30":{"pot_scale":1,"posy":62,"posx":-7.5},"24":{"pot_scale":1,"posy":62,"posx":-9.5},"23":{"pot_scale":1,"posy":257.5,"posx":-18},"22":{"pot_scale":1,"posy":62,"posx":-7.5},"17":{"pot_scale":1,"posy":254,"posx":-23},"25":{"pot_scale":1,"posy":254,"posx":-23},"6":{"posy":299.5,"posx":-23.5},"7":{"posy":376,"posx":-30.5},"8":{"posy":260,"posx":-24},"9":{"posy":146,"posx":-22.5},"29":{"pot_scale":1,"posy":258,"posx":-12.5},"28":{"pot_scale":1,"posy":66.5,"posx":-11},"27":{"pot_scale":1,"posy":238.5,"posx":-26},"26":{"pot_scale":1,"posy":73.5,"posx":-7},"16":{"pot_scale":1,"posy":62,"posx":-9.5},"19":{"pot_scale":1,"posy":238.5,"posx":-26},"18":{"pot_scale":1,"posy":73.5,"posx":-7},"1":{"posy":-29.5,"posx":-22.5},"2":{"posy":239.5,"posx":-11.5},"3":{"posy":128.5,"posx":-21},"4":{"posy":253.5,"posx":-11.5},"5":{"posy":56,"posx":-9.5},"13":{"posy":253.5,"posx":-11.5},"12":{"posy":205.5,"posx":-24},"15":{"pot_scale":1,"posy":257.5,"posx":-18},"14":{"posy":56,"posx":-9.5},"10":{"posy":-204.5,"posx":-9.5},"11":{"posy":417.5,"posx":-21.5},"21":{"pot_scale":1,"posy":258,"posx":-12.5},"20":{"pot_scale":1,"posy":66.5,"posx":-11}},"id":1}