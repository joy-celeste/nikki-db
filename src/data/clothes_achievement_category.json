{"8":"Troupe","9":"4 Seasons","5":"Others","4":"Styling Contest","7":"Festivals","6":"7 Nations","3":"Journey","21":"Museum","20":"御苑琼芳","23":"Constellation Hall","22":"Ancient Fossil Hall","15":"North","14":"Pigeon","17":"Ruin","16":"Wasteland","18":"Happiness","19":"Story Suit","2":"Collection","1":"My Growth","12":"Lilith","13":"Cloud","10":"Stars","11":"Apple"}