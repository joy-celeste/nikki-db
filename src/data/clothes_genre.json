{"Coat":3,"AccessoryCount":8,"AccessoryEnd":17,"Earring":11,"Shoes":7,"Necklace":12,"Count":10,"Top":4,"Special":16,"Bottom":5,"Hair":1,"Dress":2,"Headwear":10,"Accessory":8,"AccessoryStart":10,"Waistdeco":15,"Bracelet":13,"Socks":6,"Skin":17,"MiracleSoul":88,"Makeup":9,"Handheld":14}