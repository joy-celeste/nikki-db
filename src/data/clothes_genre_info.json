{"1":{"combat_param":0.5,"display_name":"Hair","name":"发型","sub_type":[1],"file_name":"faxing"},"2":{"combat_param":2,"display_name":"Dress","name":"连衣裙","sub_type":[2],"file_name":"lianyiqun"},"3":{"combat_param":0.2,"display_name":"Coat","name":"外套","sub_type":[3],"file_name":"waitao"},"4":{"combat_param":1,"display_name":"Top","name":"上衣","sub_type":[4],"file_name":"shangyi"},"5":{"combat_param":1,"display_name":"Bottom","name":"下装","sub_type":[5],"file_name":"xiazhuang"},"6":{"combat_param":0.3,"display_name":"Hosiery","name":"袜子","sub_type":[6,7],"file_name":"wazi"},"7":{"combat_param":0.4,"display_name":"Shoes","name":"鞋子","sub_type":[8],"file_name":"xiezi"},"8":{"display_name":"Accessory","name":"饰品","file_name":"shipin","combat_param":0.2},"9":{"combat_param":0.1,"display_name":"Makeup","name":"妆容","sub_type":[9],"file_name":"zhuangrong"},"10":{"combat_param":0.2,"display_name":"Headwear","name":"头饰","sub_type":[10,20,28,29],"file_name":"toushi"},"11":{"combat_param":0.2,"display_name":"Earrings","name":"耳饰","sub_type":[11],"file_name":"ershi"},"12":{"combat_param":0.2,"display_name":"Necklace","name":"颈饰","sub_type":[12,13],"file_name":"jingshi"},"13":{"combat_param":0.2,"display_name":"Hands","name":"手饰","sub_type":[14,15,16],"file_name":"shoushi"},"14":{"combat_param":0.2,"display_name":"Handheld","name":"手持物","sub_type":[17,18,33],"file_name":"shouchiwu"},"15":{"combat_param":0.2,"display_name":"Waist","name":"腰饰","sub_type":[19],"file_name":"yaoshi"},"16":{"combat_param":0.2,"display_name":"Special","name":"特殊","sub_type":[21,22,23,24,25,26,27,30,31,35],"file_name":"teshu"},"17":{"combat_param":0.2,"display_name":"Skin","name":"皮肤","sub_type":[32],"file_name":"pifu"},"88":{"combat_param":0.2,"display_name":"Spirit","name":"萤光之灵","sub_type":[34],"file_name":"yingguangzhiling"}}