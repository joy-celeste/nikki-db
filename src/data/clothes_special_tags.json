{"47":"Republic of China","49":"Hindu","51":"Army","41":"Bohemia","31":"Floral","21":"Pajama","23":"Britain","13":"Homewear","25":"Dryad","15":"Office","11":"Sports","17":"Unisex","29":"Rain","19":"European","27":"Shower","46":"Wedding","48":"Cheongsam","54":"Harajuku","53":"Future","52":"Navy","40":"Winter","50":"Traditional","20":"Denim","30":"Swimsuit","12":"Pop","22":"Dancer","32":"Sun Care","24":"Rock","45":"Lolita","44":"Kimono","43":"Modern China","42":"Swordsman","33":"Apron","34":"Paramedics","39":"Maiden","38":"Lady","37":"Gothic","36":"Bunny","35":"Evening Gown","14":"Chinese Classical","28":"Pet","16":"Preppy","26":"Goddess","18":"Fairy"}