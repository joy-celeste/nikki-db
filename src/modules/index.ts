import { combineReducers } from 'redux';
import { userReducer } from './user';
import { productsReducer } from './products';

export const rootReducer = combineReducers({
  user: userReducer,
  products: productsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;