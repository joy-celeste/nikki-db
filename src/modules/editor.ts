import { Dispatch } from 'react';
import { AnyAction } from 'redux';
import { RootState } from '.';
import { ACTION_CONSTANTS } from './constants';
import { ItemId } from './data';

const DEFAULT_BACKGROUND_IMAGE_NAME = 'medium';

export interface MenuState {
  closet: boolean,
  inventory: boolean,
}

export type EditorState = {
  hiddenItems: Set<ItemId>,
  backgroundImageName: string,
  downloadName: string;
  downloaded: Set<ItemId>;
  minimizedMenus: MenuState;
  activeMenus: MenuState;
};

const initialState: EditorState = {
  hiddenItems: new Set(),
  backgroundImageName: DEFAULT_BACKGROUND_IMAGE_NAME,
  downloadName: 'nikki',
  downloaded: new Set<ItemId>(),
  minimizedMenus: { closet: false, inventory: false },
  activeMenus: { closet: false, inventory: true },
};

// ACTIONS
export const setMinimized = (minimizedMenus: MenuState): AnyAction => ({
  type: ACTION_CONSTANTS.EDITOR_CHANGE_MINIMIZED_MENUS,
  payload: minimizedMenus,
});

export const setActive = (activeMenus: MenuState): AnyAction => ({
  type: ACTION_CONSTANTS.EDITOR_CHANGE_ACTIVE_MENUS,
  payload: activeMenus,
});

export const changeHiddenItemList = (hiddenItemList: Set<ItemId>): AnyAction => ({
  type: ACTION_CONSTANTS.EDITOR_CHANGE_HIDDEN_ITEM_LIST,
  payload: hiddenItemList,
});

export const setDownloadName = (downloadName: string): AnyAction => ({
  type: ACTION_CONSTANTS.EDITOR_SET_DOWNLOAD_NAME,
  payload: downloadName,
});

export const setDownloadedItems = (downloaded: Set<ItemId>): AnyAction => ({
  type: ACTION_CONSTANTS.EDITOR_SET_ALREADY_DOWNLOADED_ITEM,
  payload: downloaded,
});

// USE-CASE
export const toggleItemVisibility = (itemId: ItemId) =>
  async(dispatch: Dispatch<AnyAction>, getState: () => RootState): Promise<void> => {
    const oldHiddenItems: Set<ItemId> = getState().editor.hiddenItems;
    const newHiddenItems: Set<ItemId> = new Set(oldHiddenItems);

    if (!oldHiddenItems.has(itemId)) {
      newHiddenItems.add(itemId);
    } else {
      newHiddenItems.delete(itemId);
    }
    dispatch(changeHiddenItemList(newHiddenItems));
  };

// REDUCER
export function editorReducer(
  state = initialState,
  action: AnyAction,
): EditorState {
  switch (action.type) {
    case ACTION_CONSTANTS.EDITOR_CHANGE_HIDDEN_ITEM_LIST:
      return {
        ...state,
        hiddenItems: action.payload,
      };
    case ACTION_CONSTANTS.EDITOR_SET_DOWNLOAD_NAME:
      return {
        ...state,
        downloadName: action.payload,
      };
    case ACTION_CONSTANTS.EDITOR_SET_ALREADY_DOWNLOADED_ITEM:
      return {
        ...state,
        downloaded: action.payload,
      };
    case ACTION_CONSTANTS.EDITOR_CHANGE_MINIMIZED_MENUS:
      return {
        ...state,
        minimizedMenus: action.payload,
      };
    case ACTION_CONSTANTS.EDITOR_CHANGE_ACTIVE_MENUS:
      return {
        ...state,
        activeMenus: action.payload,
      };
    default:
      return state;
  }
}
