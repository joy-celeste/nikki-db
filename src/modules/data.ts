import { AnyAction } from 'redux';
import { DocumentData } from '@firebase/firestore-types';
import { DeserializeNullException } from './errors';
import { ACTION_CONSTANTS, BODY, BODY_ITEM_DATA, BODY_ITEM_ID, BODY_PARTS_DEPTHS,
  DEFAULT_AMPUTATIONS_LIST, DEPTHTYPE_TO_SUBTYPES, SUBTYPES_LIST, CLOTHES_DATA } from './constants';
import { Character, CharacterState, wearItem, addToHistory } from './character';
import { RootState } from '.';
import { changeHiddenItemList } from './editor';

export type SubType = typeof SUBTYPES_LIST[number];
export type AmputationParts = typeof DEFAULT_AMPUTATIONS_LIST[number];
export type AmputationData = Record<AmputationParts, boolean>;
export type Depths = Record<number, number>;
export type ItemId = number;
export type DepthType = number;

export interface SubtypeInfo {
  subtype: number,
  depth: Record<number, number>
}

export class ItemData {
  id?: ItemId;
  name?: string;
  description?: string;
  depthType?: DepthType;
  subType?: SubType;
  amputationData?: AmputationData;
  position?: PositionData[];
  depths?: Depths;

  constructor(input?: DocumentData) {
    this.deserialize(input);
  }

  deserialize(input: DocumentData): ItemData {
    if (!input) {
      throw new DeserializeNullException('Cannot deserialize null input for ItemData');
    }

    this.id = input.id || null;
    this.name = input.name || '';
    this.description = input.desc || '';
    this.position = null;
    this.depthType = null;
    this.subType = null;
    this.depths = null;
    this.amputationData = null;

    if (input.amputation_data) {
      this.amputationData = {
        [BODY.ARM]: !!input.amputation_data.arm,
        [BODY.TORSO]: !!input.amputation_data.body,
        [BODY.BREAST]: !!input.amputation_data.body,
        [BODY.LEG]: !!input.amputation_data.leg,
        [BODY.PANTY]: !!input.amputation_data.panty,
      };
    }

    if (input.depth_type || input.id === BODY_ITEM_ID) {
      this.depthType = input.depth_type;
      const subtypeData = input.id !== BODY_ITEM_ID
        ? DEPTHTYPE_TO_SUBTYPES[input.depth_type.toString()]
        : BODY_PARTS_DEPTHS;
      this.subType = subtypeData.subtype;
      this.depths = subtypeData.depth;
    }

    if (input.position) {
      const position: PositionData[] = [];
      Object.entries(input.position).forEach(([key, value]) => {
        position.push(new PositionData(parseInt(key, 10), value, this.depths));
      });
      this.position = position;
    }
    return this;
  }
}

export class PositionData {
  depth: number;
  x: number;
  y: number;
  z: number;
  height?: number;
  width?: number;
  scale?: number;

  constructor(depth: number, input: DocumentData, depths?: Depths) {
    this.deserialize(depth, input, depths);
  }

  deserialize(depth: number, input: DocumentData, depths?: Depths): PositionData {
    if (!depth || !input) {
      throw new DeserializeNullException('Cannot deserialize null input for PositionData');
    }

    this.depth = depth;
    this.x = input.posx;
    this.y = input.posy;
    this.z = null;
    this.height = input.height || null;
    this.width = input.width || null;
    this.scale = input.pot_scale || null;

    if (depths) {
      this.calculateDepth(depth, depths);
    }
    return this;
  }

  // Don't ask me about this logic... I'm just copying this straight from the original devs
  calculateDepth(depth: number, depths: Depths): void {
    if (depth < 100) {
      this.z = depths[depth];
    } else {
      const key: number = depth / 100;
      const ceil: number = Math.ceil(key);
      /* istanbul ignore next */
      if (key <= 0 || ceil === key) { // These are for animated effects.
        this.z = depths[ceil] - 5;
      }
      this.z = depths[ceil - 1] - 5;
    }
  }
}

export const bodyItemData = new ItemData(BODY_ITEM_DATA);

export type ItemsData = Record<ItemId, ItemData>;
export type DataState = {
  itemsData: ItemsData;
  loading: boolean;
};

export const initialState: DataState = {
  itemsData: { 10001: new ItemData(CLOTHES_DATA[10001]) },
  loading: false,
};

// ACTIONS
export const addItemData = (itemId: number, itemData: ItemData): AnyAction => ({
  type: ACTION_CONSTANTS.DATA_ADD_ITEMS,
  payload: {
    itemId, itemData,
  },
});

export type BreakdownData = {
  variant: string;
  itemId: ItemId;
  prosthesis: boolean;
  subtype: SubType;
  name: string;
};

// USE-CASE
export const loadItem = (itemId: ItemId) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    const items: ItemsData = getState().data.itemsData;
    if (!(itemId in items) && CLOTHES_DATA[itemId]) {
      dispatch(addItemData(itemId, new ItemData(CLOTHES_DATA[itemId])));
    } else if (itemId in items) {
      dispatch(wearItem(itemId));
    }
  };

export const loadMultipleItems = (itemIds: ItemId[]) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    dispatch(changeHiddenItemList(new Set<ItemId>()));
    if (itemIds.length === 1) {
      dispatch(loadItem(itemIds[0]));
      return;
    }

    const items: ItemsData = getState().data.itemsData;
    const charState: CharacterState = getState().character;
    const newChar: Character = new Character();
    let itemData: ItemData;

    await Promise.allSettled(itemIds.map(async (itemId) => {
      if (!(itemId in items)) {
        itemData = new ItemData(CLOTHES_DATA[itemId]);
        dispatch(addItemData(itemId, itemData));
      } else {
        itemData = items[itemId];
      }
      newChar.wear(itemData);
    })).then(() => {
      dispatch(addToHistory(newChar, charState.step + 1));
    });
  };

// REDUCER
export function dataReducer(
  state = initialState,
  action: AnyAction,
): DataState {
  switch (action.type) {
    case ACTION_CONSTANTS.DATA_ADD_ITEMS:
      return {
        ...state,
        itemsData: {
          ...state.itemsData,
          [action.payload.itemId]: action.payload.itemData,
        },
      };
    default:
      return state;
  }
}
