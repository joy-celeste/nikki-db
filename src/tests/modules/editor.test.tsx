import { combineReducers, Store } from 'redux';
import { simpleDress, simpleHair } from '../test_data/data';
import { ItemId } from '../../modules/data';
import { createStoreWithMiddleware } from '../helpers';
import { EditorState, editorReducer, toggleItemVisibility, setDownloadName, setDownloadedItems, setMinimized, setActive } from '../../modules/editor';
import { RootState } from '../../modules';

describe('EditorState', () => {
  let store: Store<any>;
  let mockRootReducer: any;
  let mockCallback: any;
  
  beforeEach(() => {
    mockRootReducer = combineReducers({
      editor: editorReducer,
    });
    store = createStoreWithMiddleware(mockRootReducer);
    mockCallback = jest.fn(input => input);
  });

  test('Assert initial state uses default values', async () => {
    const editorState: EditorState = store.getState().editor;
    expect(editorState.hiddenItems).toEqual(new Set<ItemId>());
  });

  test('Action: EDITOR_CHANGE_HIDDEN_ITEM_LIST / Use-case: toggleItemVisibility -- '
       + 'successfully hide and unhide items', async () => {
    await store.dispatch<any>(toggleItemVisibility(simpleHair.id));
    let state: RootState = store.getState();
    expect(state.editor.hiddenItems.has(simpleHair.id));
    expect(state).toMatchSnapshot();

    await store.dispatch<any>(toggleItemVisibility(simpleDress.id));
    expect(state.editor.hiddenItems.has(simpleDress.id));
    state = store.getState();
    expect(state).toMatchSnapshot();

    await store.dispatch<any>(toggleItemVisibility(simpleHair.id));
    expect(!state.editor.hiddenItems.has(simpleHair.id));
    state = store.getState();
    expect(state).toMatchSnapshot();
  });

  test('Action: EDITOR_SET_DOWNLOAD_NAME', async () => {
    await store.dispatch<any>(setDownloadName("blah"));
    let state: RootState = store.getState();
    expect(state.editor.downloadName).toEqual("blah")
  });

  test('Action: EDITOR_SET_ALREADY_DOWNLOADED_ITEM', async () => {
    const downloadedItems = new Set<ItemId>([1, 2]);
    await store.dispatch<any>(setDownloadedItems(downloadedItems));
    let state: RootState = store.getState();
    expect(state.editor.downloaded).toEqual(downloadedItems)
  });

  test('Action: EDITOR_CHANGE_MINIMIZED_MENUS', async () => {
    let state: RootState = store.getState();
    expect(state.editor.minimizedMenus.closet).toEqual(false);
    expect(state.editor.minimizedMenus.inventory).toEqual(false);
    
    await store.dispatch<any>(setMinimized({ closet: true, inventory: false }));
    state = store.getState();
    expect(state.editor.minimizedMenus.closet).toEqual(true);
    expect(state.editor.minimizedMenus.inventory).toEqual(false);

    await store.dispatch<any>(setMinimized({ closet: false, inventory: true }));
    state = store.getState();
    expect(state.editor.minimizedMenus.closet).toEqual(false);
    expect(state.editor.minimizedMenus.inventory).toEqual(true);

    await store.dispatch<any>(setMinimized({ closet: true, inventory: true }));
    state = store.getState();
    expect(state.editor.minimizedMenus.closet).toEqual(true);
    expect(state.editor.minimizedMenus.inventory).toEqual(true);
  });

  test('Action: EDITOR_CHANGE_ACTIVE_MENUS', async () => {
    let state: RootState = store.getState();
    expect(state.editor.activeMenus.closet).toEqual(false);
    expect(state.editor.activeMenus.inventory).toEqual(true);
    
    await store.dispatch<any>(setActive({ closet: true, inventory: false }));
    state = store.getState();
    expect(state.editor.activeMenus.closet).toEqual(true);
    expect(state.editor.activeMenus.inventory).toEqual(false);

    await store.dispatch<any>(setActive({ closet: false, inventory: false }));
    state = store.getState();
    expect(state.editor.activeMenus.closet).toEqual(false);
    expect(state.editor.activeMenus.inventory).toEqual(false);

    await store.dispatch<any>(setActive({ closet: true, inventory: true }));
    state = store.getState();
    expect(state.editor.activeMenus.closet).toEqual(true);
    expect(state.editor.activeMenus.inventory).toEqual(true);
  });
});
