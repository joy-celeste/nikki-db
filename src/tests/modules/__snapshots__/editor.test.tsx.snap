// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorState Action: EDITOR_CHANGE_HIDDEN_ITEM_LIST / Use-case: toggleItemVisibility -- successfully hide and unhide items 1`] = `
Object {
  "editor": Object {
    "activeMenus": Object {
      "closet": false,
      "inventory": true,
    },
    "backgroundImageName": "medium",
    "downloadName": "nikki",
    "downloaded": Set {},
    "hiddenItems": Set {
      10007,
    },
    "minimizedMenus": Object {
      "closet": false,
      "inventory": false,
    },
  },
}
`;

exports[`EditorState Action: EDITOR_CHANGE_HIDDEN_ITEM_LIST / Use-case: toggleItemVisibility -- successfully hide and unhide items 2`] = `
Object {
  "editor": Object {
    "activeMenus": Object {
      "closet": false,
      "inventory": true,
    },
    "backgroundImageName": "medium",
    "downloadName": "nikki",
    "downloaded": Set {},
    "hiddenItems": Set {
      10007,
      21305,
    },
    "minimizedMenus": Object {
      "closet": false,
      "inventory": false,
    },
  },
}
`;

exports[`EditorState Action: EDITOR_CHANGE_HIDDEN_ITEM_LIST / Use-case: toggleItemVisibility -- successfully hide and unhide items 3`] = `
Object {
  "editor": Object {
    "activeMenus": Object {
      "closet": false,
      "inventory": true,
    },
    "backgroundImageName": "medium",
    "downloadName": "nikki",
    "downloaded": Set {},
    "hiddenItems": Set {
      21305,
    },
    "minimizedMenus": Object {
      "closet": false,
      "inventory": false,
    },
  },
}
`;
